torch_xla/experimental/spmd_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,107 @@
+    return partition_spec
+
+  # Only shard the 0th dimension of the parameter according to the
+  # fsdp axis of the mesh.
+  # TODO: should we shard on the maximal dim for param? Then we need
+  # another helper for the output.
+  partition_spec[0] = "fsdp"
+  return tuple(partition_spec)
+
+
+class SpmdFullyShardedDataParallel(nn.Module):
+
+  def __init__(self, module: nn.Module, mesh: spmd.Mesh, shard_output:Optional[Callable] = None):
+    if isinstance(module, SpmdFullyShardedDataParallel):
+      raise RuntimeError(
+          "Cannot wrap a module that is already wrapped with FSDP. For nested FSDP, "
+          "first wrap the inner child modules before wrapping the outer parent module."
+      )
+    is_forward_defined = (
+        hasattr(module, "forward") and hasattr(module.forward, "__func__") and
+        module.forward.__func__ != nn.Module.forward)
+    if not is_forward_defined:
+      raise RuntimeError(
+          "The module wrapped by FSDP *must define a `forward` method and call it "
+          "during the module's forward pass for FSDP to work correctly.* "
+          "Hence, do not wrap `nn.ModuleList` or `nn.ModuleDict` with FSDP "
+          "(since they don't have `forward` defined), "
+          "and do not perform the forward pass in other ways apart from the `forward` method. "
+          "(i.e. you should directly call the FSDP-wrapped module itself in your code, "
+          "instead of using any of its submodules or its weights).")
+    if "fsdp" not in mesh.axis_names:
+      raise ValueError("The mesh must have an axis named 'fsdp'.")
+
+    super().__init__()
+
+    self._orig_module = module
+    self._mesh = mesh
+
+    # Only handle params which are not already sharded. This enables
+    # sharding individual layers of a Module, with an outer wrapper to
+    # shard any leftover parameters.
+    for param in module.parameters():
+      if torch_xla._XLAC._get_xla_sharding_spec(param) != "":
+        continue
+      spmd.mark_sharding(param, mesh, _prepare_spmd_partition_spec(param))
+
+    # Register a backward hook to place optimization barrier to prevent
+    # gigantic fusions on syncing the gradients.
+    spmd.xla_sharding.apply_backward_optimization_barrier(module)
+
+    # Need to shard the output of the forward to instruct the compiler
+    # to enforce the FSDP algorithm.
+    if shard_output is None:
+      def shard_output_impl(output, mesh):
+        real_output = None
+        if isinstance(output, TensorLike):
+          real_output = output
+        # TODO: enhance the logic as real_output can be something else. use _extract_tensors_from_list
+        elif isinstance(output, tuple):
+          real_output = output[0]
+        else:
+          raise RuntimeError(
+              f"The output type is not supported: {type(output)}. Please provide your own shard_output callable.")
+
+        spmd.mark_sharding(real_output, mesh, _prepare_spmd_partition_spec(real_output))
```

Comment:
It looks like we can use SPMDFullyShardedDataParallel to express data-parallel multislice by specifying a mesh like `('dcn', 'fsdp')`, but we'd need to combine the two axes when sharding the activations' batch axis.

We could achieve this by specifying a `shard_output` function, but what do you think of allowing users to override which axes to shard activations along in the default `shard_output_impl`? e.g. a new constructor parameter for `activation_sharding='fsdp'`, and allow it to be overridden to `activation_sharding=('dcn', 'fsdp')`

Reply:
Oops, I missed that.

Reply:
Let's do this as a follow up.